### DISCUSSION TITLE
Episode 124: Vectorizing std::views::filter

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#15] [Episode]
Episode Link: https://adspthepodcast.com/2023/04/07/Episode-124.html

Feel free to write comments or ask questions 🙂

**Comment by tmr232** [comment]
To get code that's easy to see in the assembly, you can use function calls instead of "simple operations".
As long as you don't provide function bodies, they won't be inlined, and you'll always get their names.

https://godbolt.org/z/9Wb7M7Wda

**Reply by roibarkan** [reply]
yeah, but function calls which aren't inlined will not be vectorizable. I believe Bryce is attempting to create code that the compiler understands enough to vectorize

**Comment by roibarkan** [comment]
I think the various algorithms that end with `_if` might be a good approach for doing the types of operations that Bryce was talking about.
I think something like `transform_if(range, operation, predicate)` would be a good approach for this specific case.
Potentially one can also `zip` a boolean with the input range in order to perhaps be more composable (very similar overhead as converting to `optional` but with the known `zip` terminology).
So one can imagine something like `range | zip_with_func(predicate) | transform([](auto pair) {return pair.second? Op(pair.first) : nulopt;})` to achieve what Bryce intended.

**Comment by axman6** [comment]
Sounds like another re-invention of (vectorised) stream fusion - https://www.microsoft.com/en-us/research/wp-content/uploads/2016/07/haskell-beats-C.pdf?from=https://research.microsoft.com/en-us/um/people/simonpj/papers/ndp/haskell-beats-C.pdf&type=exact but, as always with C++, it’s been over complicated by people trying to figure out every way someone might want to use something instead of thinking of the smallest possible parts that could be used to implement everything.
